rfcs/0042-config-option.md

```diff
@@ -0,0 +1,259 @@
+
+```nix
+{ config, pkgs, lib, ... }: with lib;
+
+let
+
+  configText = configGen.json cfg.config;
+
+  configFile = pkgs.runCommand "foo-config.json" {
+    # Because this program will be run at build time, we need `nativeBuildInputs` instead of `buildInputs`
+    nativeBuildInputs = [ pkgs.foo ];
+
+    inherit configText;
+    passAsFile = [ "configText" ];
+  } ''
+    foo check-config $configTextPath
+    cp $configTextPath $out
+  '';
+
+in { /* ... */ }
+```
+
+#### Ad-hoc checks with assertions
+
+While not as optimal as a configuration checker tool, assertions can be used to add flexible ad-hoc checks for type or other properties at **evaluation time**. It should only be used to ensure important properties that break the service in ways that are otherwise hard or slow to detect (and easy to detect for the module system), not for things that make the service fail to start anyways (unless there's a good reason for it). The following example only demonstrates how assertions can be used for checks, but any reasonable program should bail out early in such cases, which would make these assertions redundant, and only add more coupling to upstream, which we're trying to avoid.
+
+```nix
+{ config, lib, ... }: with lib; {
+  # ...
+  config = mkIf cfg.enable {
+    # Examples only for demonstration purposes, don't actually add assertions for such properties
+    assertions = [
+      {
+        assertion = cfg.config.enableLogging or true -> cfg.config ? logLevel;
+        message = "You also need to set `services.foo.config.logLevel` if `services.foo.config.enableLogging` is turned on.";
+      }
+      {
+        assertion = cfg.config ? port -> types.port.check cfg.config.port;
+        message = "${toString cfg.config.port} is not a valid port number for `services.foo.config.port`.";
+      }
+    ];
+  };
+}
+```
+
+TODO: Are there any good examples of using assertions for configuration checks at all?
+
+### Implementation
+
+The implementation consists of three separate parts
+
+#### Configuration types
+
+A set of types for common configuration formats should be provided in `lib.types.config`. Such a type should encode what values can be set in files of this configuration format as a Nix value, with the module system being able to merge multiple values correctly. This is the part that checks whether the user set an encodeable value. This can be extended over time, but could include the following as a start:
+- JSON
+- YAML, which is probably the same as JSON
+- INI
+- A simple `key=value` format
+- A recursive `key.subkey.subsubkey=value` format
```

Comment:
Please note, that not every configuration file format perfectly fits into that. I'm doing what you propose since years (keeping a `.config` in addition to higher-level options that set certain values) and I can remember several occasions where it's a bit more tricky to do, eg. with things like eg.:

```ini
a=foo
a.b=bar
```

So I'd probably add some convention around something like this, eg.:

```nix
{
  a.__toplevel = "foo";
  a.b = "bar";
}
```

Not sure however if the following paragraph addresses that (eg. what do you mean with `should be implemented in the program's module directly`?).

Reply:
Well with the example you mentioned it would make sense for it to just be
```nix
{
  a = "foo";
  a.b = "bar";
}
```
This would be implemented recursively. But for INI such a `__toplevel`/`__global` attribute might be needed for global attributes without a section.

Reply:
By `should be implemented in the program's module directly` I mean that it doesn't need to have its own library function in `nixpkgs/lib` when it only ever gets used for a single program.

Reply:
the `a.__toplevel` is needed to avoid Nix confusion: is `a` a string or attrset?

Reply:
Oh I get it now, what I showed is invalid Nix haha. But I don't get the point being made, what configuration allows something to be a string and an attrset at the same time? This doesn't make any sense to me.

Reply:
The example you mentioned is invalid Nix... However some languages like eg. PHP do something like this (eg. see the [List of pool directives](https://www.php.net/manual/en/install.fpm.configuration.php)). However as discussed on IRC, given that INI files aren't recursive, we can just use something like `{ a = "foo"; "a.b" = "bar"; }`, which should work fine.